x-pack/plugins/lens/server/migrations.test.ts

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { migrations, RawLensSavedXYObject770 } from './migrations';
+import { SavedObjectUnsanitizedDoc } from 'src/core/server';
+
+describe('Lens migrations', () => {
+  describe('7.7.0 missing dimensions in XY', () => {
+    const migrate = (doc: SavedObjectUnsanitizedDoc | RawLensSavedXYObject770) =>
+      migrations['7.7.0'](doc);
+
```

Comment:
Why not just `const migrate = migrations['7.7.0']` ?

On a higher level, that kind of specialized input type makes sense imho.

Maybe we should adapt

https://github.com/elastic/kibana/blob/8524303b6e5cac434353d57700a31e907ab93f5a/src/core/server/saved_objects/migrations/types.ts#L42-L45

to

```ts
export type SavedObjectMigrationFn = <
  InputDocDocType extends SavedObjectUnsanitizedDoc = SavedObjectUnsanitizedDoc,
  MigratedDocType extends SavedObjectUnsanitizedDoc = SavedObjectUnsanitizedDoc
>(
  doc: InputDocDocType,
  context: SavedObjectMigrationContext
) => MigratedDocType;
```

@rudolf WDYT?

Also I think you already stated in another PR that we did not really want to export the `SavedObjectUnsanitizedDoc` type. Can you confirm or infirm that?

Reply:
Okay, I'm going to get rid of these type checks.

Reply:
> A migration can only ever receive documents of it's type (@pgayvallet correct me if I'm wrong), 

You are right (...)

> so this is actually a bug in core's types or just our types being so loose they're misleading.

The main issue regarding typings here is that there is no TS link between a `SavedObjectsType` (or even `SavedObjectsType.mappings`) and it's associated object's properties (`SavedObject<Properties>`) and even less with the raw type we are using for migrations `SavedObjectUnsanitizedDoc`, meaning that there is ATM no way to automatically infer the types for the migration function input or output. 

This is why adding generic types to `SavedObjectMigrationFn` sounds to me like a quick win to allow plugin developers to specify the input/output without having to manually redefine the signature.